File: chapter_03/recursive_binary_search.py
def recursive_binary_search(lst, item):
    if len(lst) == 0:
        return None
    
    mid = len(lst) // 2
    guess = lst[mid]

    if guess == item:
        return mid
    
    if guess > item:
        return recursive_binary_search(lst[:mid], item)
    
    else:
        return recursive_binary_search(lst[mid + 1:], item)

print(recursive_binary_search([1, 3, 5, 7, 9, 14 ,15 ,16 ,17 ,18 ,19 ,20], 15))

File: chapter_03/RecursiveBinarySearch.cs
using System;

namespace Algorithms.Chapter03
{
    class RecursiveBinarySearchProgram
    {
        public static int? RecursiveBinarySearch(int[] lst, int item)
        {
            if (lst.Length == 0)
            {
                return null;
            }

            int mid = lst.Length / 2;
            int guess = lst[mid];

            if (guess == item)
            {
                return mid;
            }

            if (guess > item)
            {
                return RecursiveBinarySearch(lst[..mid], item);
            }
            else
            {
                return RecursiveBinarySearch(lst[(mid + 1)..], item);
            }
        }

        public static void Main()
        {
            int[] arr = { 1, 3, 5, 7, 9, 14, 15, 16, 17, 18, 19, 20 };
            Console.WriteLine(RecursiveBinarySearch(arr, 15));
        }
    }
}
